be/src/olap/tablet.h

```diff
@@ -403,20 +405,45 @@ class Tablet : public BaseTablet {
                            OlapReaderStatistics& stats, vectorized::Block* block,
                            bool write_to_cache = false);
 
+    Status fetch_value_by_rowids(RowsetSharedPtr input_rowset, uint32_t segid,
+                                 const std::vector<uint32_t>& rowids,
+                                 const std::string& column_name, vectorized::MutableColumnPtr& dst);
+
+    Status fetch_value_through_row_column(RowsetSharedPtr input_rowset, uint32_t segid,
+                                          const std::vector<uint32_t>& rowids,
+                                          const std::vector<uint32_t>& cids,
+                                          vectorized::Block& block);
+
     // calc delete bitmap when flush memtable, use a fake version to calc
     // For example, cur max version is 5, and we use version 6 to calc but
     // finally this rowset publish version with 8, we should make up data
     // for rowset 6-7. Also, if a compaction happens between commit_txn and
     // publish_txn, we should remove compaction input rowsets' delete_bitmap
     // and build newly generated rowset's delete_bitmap
-    Status calc_delete_bitmap(RowsetId rowset_id,
+    Status calc_delete_bitmap(RowsetSharedPtr rowset,
                               const std::vector<segment_v2::SegmentSharedPtr>& segments,
                               const RowsetIdUnorderedSet* specified_rowset_ids,
                               DeleteBitmapPtr delete_bitmap, int64_t version,
-                              bool check_pre_segments = false);
+                              bool check_pre_segments = false,
+                              RowsetWriter* rowset_writer = nullptr);
+    Status read_columns_by_plan(
+            TabletSchemaSPtr tablet_schema, const std::vector<uint32_t> cids_to_read,
+            const std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>>& read_plan,
+            const std::map<RowsetId, RowsetSharedPtr>& rsid_to_rowset, vectorized::Block& block,
+            std::map<uint32_t, uint32_t>* read_index);
+    void prepare_to_read(const RowLocation& row_location, size_t pos,
+                         std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>>* read_plan);
+    Status generate_new_block_for_partial_update(
+            TabletSchemaSPtr rowset_schema,
+            const std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>>& read_plan_ori,
```

Comment:
consider to use a type alias for `std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>>` to make the code clear, e.g.

typedef std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>> ReadPlan

Reply:
fixed